### PR TITLE
UCP/RMA/NBX: added implementation of NBX API

### PR DESCRIPTION
**Pull request by hoopoepg**
- added implementation of get/put _nbx calls

### CODE REVIEW

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
@yosefe ok to squash?
will fix @brminich's comment

File: src/ucp/rma/rma.inl [hunk 1]
```diff
@@ -32,6 +33,24 @@ ucp_rma_send_request_cb(ucp_request_t *req, ucp_send_callback_t cb)
...
+    ucs_status_t status = ucp_request_send(req, 0);
+
+    if (req->flags & UCP_REQUEST_FLAG_COMPLETED) {
+        ucp_request_imm_cmpl_param(param, req, status, send);
+    }
+
+    ucs_trace_req("returning request %p, status %s", req,
+                  ucs_status_string(status));
+    if (param->op_attr_mask & UCP_OP_ATTR_FIELD_CALLBACK) {
+        ucp_request_set_callback(req, send.cb, param->cb.send, param->user_data);
```
Comment (yosefe) [inline src/ucp/rma/rma.inl:48]: user_data flag is not checked
can we unite code with tag send?

File: src/ucp/rma/rma_send.c [hunk 2]
```diff
@@ -149,91 +149,59 @@ ucp_rma_request_init(ucp_request_t *req, ucp_ep_h ep, const void *buffer,
...
 
 ucs_status_t ucp_put_nbi(ucp_ep_h ep, const void *buffer, size_t length,
                          uint64_t remote_addr, ucp_rkey_h rkey)
 {
-    ucp_ep_rma_config_t *rma_config;
-    ucs_status_t status;
-
-    UCP_RMA_CHECK(ep->worker->context, buffer, length);
+    ucp_request_param_t param = {.op_attr_mask = 0};
+    ucs_status_ptr_t status;
```
Comment (yosefe) [inline src/ucp/rma/rma_send.c:177]: rename status -> status_ptr

File: src/ucp/rma/rma_send.c [hunk 3]
```diff
@@ -261,61 +232,68 @@ ucs_status_ptr_t ucp_put_nb(ucp_ep_h ep, const void *buffer, size_t length,
...
-    return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
-}
-
 ucs_status_t ucp_get_nbi(ucp_ep_h ep, void *buffer, size_t length,
                          uint64_t remote_addr, ucp_rkey_h rkey)
 {
-    ucp_ep_rma_config_t *rma_config;
-    ucs_status_t status;
+    ucp_request_param_t param = {.op_attr_mask = 0};
+    ucs_status_ptr_t status;
```
Comment (yosefe) [inline src/ucp/rma/rma_send.c:253]: status -> status_ptr

File: src/ucp/rma/rma_send.c [hunk 4]
```diff
@@ -261,61 +232,68 @@ ucs_status_ptr_t ucp_put_nb(ucp_ep_h ep, const void *buffer, size_t length,
...
 }
 
 ucs_status_ptr_t ucp_get_nb(ucp_ep_h ep, void *buffer, size_t length,
                             uint64_t remote_addr, ucp_rkey_h rkey,
                             ucp_send_callback_t cb)
+{
+    ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK,
+        .cb.send      = (ucp_send_nbx_callback_t)cb,
+        .user_data    = NULL
```
Comment (yosefe) [inline src/ucp/rma/rma_send.c:272]: no need to set user_data

File: src/ucp/rma/rma_send.c [hunk 5]
```diff
@@ -149,91 +149,59 @@ ucp_rma_request_init(ucp_request_t *req, ucp_ep_h ep, const void *buffer,
...
 }
 
 ucs_status_ptr_t ucp_put_nb(ucp_ep_h ep, const void *buffer, size_t length,
                             uint64_t remote_addr, ucp_rkey_h rkey,
                             ucp_send_callback_t cb)
+{
+    ucp_request_param_t param = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK,
+        .cb.send = (ucp_send_nbx_callback_t)cb,
+        .user_data = NULL
```
Comment (yosefe) [inline src/ucp/rma/rma_send.c:196]: no need to set user_data

File: src/ucp/rma/rma.inl (hunk 1 above)
Comment (hoopoepg) [reply src/ucp/rma/rma.inl]: yep, unified

File: src/ucp/core/ucp_request.inl [hunk 6]
```diff
@@ -45,6 +45,11 @@
     (_rdesc)->payload_offset, \
     ((_rdesc)->length - (_rdesc)->payload_offset)
 
+#define UCP_MINIMAL_OP_ATTR \
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_request.inl]: removed

File: src/ucp/core/ucp_request.inl [hunk 7]
```diff
@@ -120,6 +130,15 @@
     return UCS_STATUS_PTR(_status);
 
 
+#define ucp_request_set_send_callback_param(_param, _req) \
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/core/ucp_request.inl]: yes, re-implemented
actually some other macro here may be implemented as inline functions

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

**Comment by hoopoepg** [comment]
@yosefe ok to merge?
AMO implementation depends from this PR

**Pull request merged**